content/blog/testing-kafka-connectors.asciidoc

```diff
@@ -0,0 +1,317 @@
+      .withNetwork(network);
+
+  // <2>
+  public static DebeziumContainer connectContainer = new DebeziumContainer("debezium/connect-base:1.9.5.Final")
+      .withFileSystemBind("target/kcetcd-connector", "/kafka/connect/kcetcd-connector")
+      .withNetwork(network)
+      .withKafka(kafkaContainer)
+      .dependsOn(kafkaContainer);
+
+  // <3>
+  public static EtcdContainer etcdContainer = new EtcdContainer("gcr.io/etcd-development/etcd:v3.5.4",
+      "etcd-a", Arrays.asList("etcd-a"))
+          .withNetworkAliases("etcd")
+          .withNetwork(network);
+
+  @BeforeAll
+  public static void startContainers() throws Exception {
+    createConnectorJar(); // <4>
+
+    Startables.deepStart(Stream.of(
+            kafkaContainer, etcdContainer, connectContainer))
+            .join();
+  }
+
+  @Test
+  public void shouldHandleAllTypesOfEvents() throws Exception {
+    Client client = Client.builder()
+        .endpoints(etcdContainer.clientEndpoint()).build();
+
+    // <5>
+    ConnectorConfiguration connector = ConnectorConfiguration.create()
+        .with("connector.class", "dev.morling.kcetcd.source.EtcdSourceConnector")
+        .with("clusters", "test-etcd=http://etcd:2379")
+        .with("tasks.max", "2")
+        .with("key.converter", "org.apache.kafka.connect.storage.StringConverter")
+        .with("value.converter", "org.apache.kafka.connect.storage.StringConverter");
+
+    // <6>
+    connectContainer.registerConnector("my-connector", connector);
+    connectContainer.ensureConnectorTaskState("my-connector", 0, State.RUNNING);
+
+    KV kvClient = client.getKVClient();
+
+    // <7>
+    // insert
+    ByteSequence key = ByteSequence.from("key-1".getBytes());
+    ByteSequence value = ByteSequence.from("value-1".getBytes());
+    kvClient.put(key, value).get();
+
+    // update
+    key = ByteSequence.from("key-1".getBytes());
+    value = ByteSequence.from("value-1a".getBytes());
+    kvClient.put(key, value).get();
+
+    // delete
+    key = ByteSequence.from("key-2".getBytes());
+    kvClient.delete(key).get();
+
+    // <8>
+    List<ConsumerRecord<String, String>> records = drain(getConsumer(kafkaContainer), 3);
+
+    // insert
+    ConsumerRecord<String, String> record = records.get(0);
+    assertThat(record.key()).isEqualTo("key-1");
+    assertThat(record.value()).isEqualTo("value-1");
+
+    // update
+    record = records.get(1);
+    assertThat(record.key()).isEqualTo("key-1");
+    assertThat(record.value()).isEqualTo("value-1a");
+
+    // delete
+    record = records.get(2);
+    assertThat(record.key()).isEqualTo("key-2");
+    assertThat(record.value()).isNull();
+  }
+}
+----
+<1> Set up Apache Kafka
+<2> Set up Kafka Connect, mounting the _target/kcetcd-connector_ directory onto the plug-in path; as part of the project's Maven build, all the dependencies of the kcetcd connector are copied into that directory
+<3> Set up Etcd
```

Comment:
Is this also using a Testcontainer module?

Reply:
Yes, it is; will clarify.